geode-core/src/test/java/org/apache/geode/internal/net/SocketUtilsJUnitTest.java

```diff
@@ -88,4 +95,86 @@ public void testCloseServerSocketThrowsIOException() throws IOException {
   public void testCloseServerSocketWithNull() {
     assertThat(SocketUtils.close((ServerSocket) null)).isTrue();
   }
+
+  @Test
+  public void readFromSocketWithHeapBuffer() throws IOException {
+    Socket socket = mock(Socket.class);
+    SocketChannel channel = mock(SocketChannel.class);
+    when(socket.getChannel()).thenReturn(channel);
+    final ByteBuffer buffer = ByteBuffer.allocate(100); // heap buffer
+    byte[] bytes = new byte[100];
+    InputStream stream = new ByteArrayInputStream(bytes);
+    when(channel.read(buffer)).thenAnswer((answer) -> {
+      buffer.put(bytes);
+      return buffer.position();
+    });
+    assertThat(buffer.hasArray()).isTrue();
+    SocketUtils.readFromSocket(socket, buffer, stream);
+    // the channel was used to read the bytes
+    verify(channel, times(1)).read(buffer);
+    // the buffer was filled
+    assertThat(buffer.position()).isEqualTo(bytes.length);
+    // the stream was not used
+    assertThat(stream.available()).isEqualTo(bytes.length);
+  }
+
+
+  @Test
+  public void readFromSocketWithDirectBuffer() throws IOException {
+    Socket socket = mock(Socket.class);
+    SocketChannel channel = mock(SocketChannel.class);
+    when(socket.getChannel()).thenReturn(channel);
+    final ByteBuffer buffer = ByteBuffer.allocateDirect(100); // non-heap buffer
```

Comment:
Great to see this increased code coverage!

From scanning this method and the previous one I think the only two differences are 1) the allocation statement and 2) the verification of `buffer.hasArray()`. It would aid maintenance, I think, if the parts that are identical were factored out.

Reply:
I don't like to factor out code in tests too much.  It leads to tests that are difficult to understand.  Short tests like this should be easy to read and stand on their own.

Reply:
fair.